### PR TITLE
[serve] Support for streaming responses

### PR DESCRIPTION
**Pull request by edoakes**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -65,16 +65,48 @@ async def resolve_async_tasks(self):
...
+    async def assign_replica(
+        self, query: Query
+    ) -> Union[ray.ObjectRef, "ray._raylet.StreamingObjectRefGenerator"]:
+        pass
+
+    def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+        pass
+
+
+class RandomStreamingReplicaScheduler(ReplicaScheduler):
```
Comment (edoakes) [inline python/ray/serve/_private/router.py:78]: The reason for this is we aren't able to use the same "tracker" ref solution here... going to have to rethink how we do the scheduling, maybe "power of 2" choice routing like we've discussed before

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -467,7 +499,9 @@ async def invoke_single(self, request_item: Query) -> Tuple[Any, bool]:
                     # call with non-empty args
                     result = await method_to_call(*args, **kwargs)
 
-            result = await self.ensure_serializable_response(result)
+            if should_convert_streaming_response:
```
Comment (sihanwang41) [inline python/ray/serve/_private/replica.py:502]: do we assume that chunk of stream result is always serializable?

File: python/ray/serve/_private/router.py (hunk 1 above)
Comment (sihanwang41) [reply python/ray/serve/_private/router.py:78]: I think we can just go for round robin to select replica in the router at the begining. 
Definitely need to understand the scenario more. qps might not be best factor to decide which replica to use for streaming. For some video/image use case, network & connection might be very important to consider.

Further idea: we can have some system level agent to expose metrics across cluster. And router can make some "smart" choice to decide which one to use.

**Review by sihanwang41 — COMMENTED** [review]
Overall looks very exciting! 
qq:
If user is dead, not able to stream results, are we able to clean up the StreamingObjectRef?

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -277,14 +276,19 @@ def __init__(
         controller_handle: ActorHandle,
         deployment_name: str,
         event_loop: asyncio.BaseEventLoop = None,
+        _use_ray_streaming: bool = False,
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:279]: Are we planning to create an API where a single handle can call both streaming and non-streaming methods on its deployment? The prototype currently marks the handle as entirely stream or entirely non-streaming.

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work! I think this approach makes sense at a high level.

**Comment by edoakes** [comment]
Microbenchmark results w/o and w/ streaming in HTTP proxy:
```
2,3c2,3
< 	single client small data 824.72 +- 7.67 requests/s
< 	8 clients small data 1195.15 +- 16.77 requests/s
---
> 	single client small data 751.52 +- 12.01 requests/s
> 	8 clients small data 2089.91 +- 8.0 requests/s
5,6c5,6
< 	single client small data 850.79 +- 27.28 requests/s
< 	8 clients small data 2643.47 +- 31.58 requests/s
---
> 	single client small data 761.34 +- 3.25 requests/s
> 	8 clients small data 2085.15 +- 9.33 requests/s
8,9c8,9
< 	single client small data 785.16 +- 29.52 requests/s
< 	8 clients small data 2289.95 +- 40.85 requests/s
---
> 	single client small data 723.36 +- 3.06 requests/s
> 	8 clients small data 1915.41 +- 12.92 requests/s
11,12c11,12
< 	single client small data 723.65 +- 4.32 requests/s
< 	8 clients small data 2568.35 +- 44.4 requests/s
---
> 	single client small data 669.41 +- 3.24 requests/s
> 	8 clients small data 2308.61 +- 17.68 requests/s
14,15c14,15
< 	single client small data 727.35 +- 17.57 requests/s
< 	8 clients small data 2682.27 +- 50.04 requests/s
---
> 	single client small data 673.95 +- 3.97 requests/s
> 	8 clients small data 2346.38 +- 27.88 requests/s
17,18c17,18
< 	single client small data 697.5 +- 5.09 requests/s
< 	8 clients small data 2646.26 +- 8.97 requests/s
---
> 	single client small data 630.5 +- 8.46 requests/s
> 	8 clients small data 2250.25 +- 22.06 requests/s
20,21c20,21
< 	single client small data 401.31 +- 5.33 requests/s
< 	8 clients small data 756.19 +- 12.02 requests/s
---
> 	single client small data 374.46 +- 1.53 requests/s
> 	8 clients small data 701.11 +- 7.56 requests/s
23,24c23,24
< 	single client small data 410.75 +- 1.36 requests/s
< 	8 clients small data 1161.42 +- 13.34 requests/s
---
> 	single client small data 381.2 +- 8.12 requests/s
> 	8 clients small data 1011.29 +- 9.86 requests/s
26,27c26,27
< 	single client small data 396.59 +- 4.15 requests/s
< 	8 clients small data 1155.63 +- 14.28 requests/s
---
> 	single client small data 365.81 +- 9.67 requests/s
> 	8 clients small data 1005.63 +- 9.11 requests/s
29,30c29,30
< 	single client small data 376.66 +- 8.92 requests/s
< 	8 clients small data 1107.74 +- 8.29 requests/s
---
> 	single client small data 360.89 +- 0.86 requests/s
> 	8 clients small data 968.92 +- 6.39 requests/s
32,33c32,33
< 	single client small data 379.55 +- 4.16 requests/s
< 	8 clients small data 1115.63 +- 13.55 requests/s
---
> 	single client small data 360.81 +- 3.71 requests/s
> 	8 clients small data 991.21 +- 11.48 requests/s
35,36c35,36
< 	single client small data 365.82 +- 4.04 requests/s
< 	8 clients small data 1120.84 +- 3.33 requests/s
---
> 	single client small data 349.14 +- 1.82 requests/s
> 	8 clients small data 978.97 +- 8.03 requests/s
```

**Pull request closed without merging**